Refactor Foreach Tests: Unary Functions

crcrpar opened the pull request:
Related issue: #58833

__changes__
- slowpath tests: pass every dtype&device tensors and compare the behavior with regular functions including inplace
- check of #cudaLaunchKernel
- rename `ForeachUnaryFuncInfo` -> `ForeachFuncInfo`: This change is mainly for the future binary/pointwise test refactors

cc: @ngimel @ptrblck @mcarilli

crcrpar committed:
ForeachUnaryFuncInfo -> ForeachFuncInfo

Also, avoid visual indentation

crcrpar committed:
Use inplace regular function as well

Also, get regular functions from `torch` namespace instead of
`torch.Tensor`

crcrpar committed:
add noncontiguous option for slowpath test

crcrpar committed:
refactor tests for unary functions

- check the number of `cudaLaunchKernel`s
- use `foreach_unary_op_db`

crcrpar committed:
reflect #57910

ngimel commented on test/test_foreach.py:
```diff
@@ -12,14 +12,67 @@
+        self.method_name = method_name
+
+    def __call__(self, inputs, **kwargs):
+        return [getattr(i1, self.method_name)(*rest, **kwargs) for i1, *rest in list(zip(*inputs))]
```
you should have gotten an attribute on torch.Tensor in `get_foreach_method_names`, and then call it here with `i1` as the first argument, it'll work like method.

ngimel commented on test/test_foreach.py:
```diff
@@ -127,27 +180,62 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
+        except RuntimeError as e:
+            if method.func == torch._foreach_neg and dtype == torch.bool and self.is_cuda:
+                # note(mkozuki): `foreach_neg` CUDA implementation throws a different message.
+                with self.assertRaisesRegex(type(e), "Negation, the `-` operator, on a bool tensor is not supported."):
```
Is it possible to make foreach_neg throw the same error message?

ngimel commented on test/test_foreach.py:
```diff
@@ -127,27 +180,62 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
+
     @ops(foreach_unary_op_db)
-    def test_unary(self, device, dtype, op):
+    def test_unary_fastpath(self, device, dtype, op):
```
Here you are testing that `foreach` op supports a subset of dtypes supported by its regular counterpart, but you won't catch a situation where regular op supports more dtypes, or a new type is added to regular op, but not foreach. This is not a blocker for this PR, but we should think of a way to keep OpInfos in sync.

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3683,93 +3685,121 @@ def _tensor(shape, dtype=dtype, low=None, high=None):
+    ForeachFuncInfo('sinh'),
+
+    ForeachFuncInfo(
+        'neg',
```
Here you could create a helper function `getOpInfo` that, based on a name, would return regular OpInfo for that name, and OpInfo could be queried for supported dtypes and easily get regular and inplace methods for the reference function. Then you wouldn't have to set dtypes manually.

ngimel reviewed (commented):
This generally looks good, I have a few minor comment. 
A bigger point is that we should have a better way of keeping regular OpInfos and for_each OpInfos in sync, and probably fill most of the attributes of for_each OpInfo based on the regular one, but we can think about how to do that later.

crcrpar committed:
make `noncontiguous` keyword only

crcrpar committed:
update handling of inplace reference function

crcrpar committed:
add skipMeta

crcrpar committed:
rename

- method -> op in common_methods_invocations.py
- op -> opinfo in test_foreach.py

crcrpar committed:
same error message from negation

crcrpar committed:
inplace negation error message

crcrpar committed:
generate tensors of different sizes

crcrpar commented:
Thank you for the review 😃 
I agree with the idea of synchronizing dtypes supported by regular functions & their counterpart foreach functions.
Though I feel like there's still a need to maintain the foreach supported dtypes if we keep checking the number of cuda kernel launches (= check of whether or not fastpath is expectedly called).
Correct me should I misunderstand your intention

ngimel commented:
> Eventually I'd like to foreach to support as many dtypes as counterpart regular functions but I think we need to maintain the supported dtypes manually meanwhile to check the number of cuda kernel launch's to tell whether the fastpath is chosen expectedly.

Can you elaborate on that? You expect not all of the dtypes to be supported by fast path, or is there going to be some other source of discrepancy?
I agree that in current state the supported dtype lists are probably different, so we can't use that approach right away.

ngimel commented on test/test_foreach.py:
```diff
@@ -127,27 +173,53 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
+    def _inplace_unary_test(self, dtype, inplace, inplace_ref, inputs, is_fastpath):
+        copied_inputs = [[t.clone().detach() for t in tensors] for tensors in inputs]
+        try:
+            inplace(inputs, self.is_cuda, is_fastpath)
```
out of curiosity, what op throws an error here? In regular (not inplace) unary test you don't need this try-except block. Can you please leave a comment about it in the test?

crcrpar replied:
Slowpath test uses all dtypes including ones that regular functions do not support: for example, `log` and `log2` do not support fp16 CPU tensors.
Thus, I use `try-except` block to check that the behavior of foreach slowpath is the same as that of the counterpart regular function.

crcrpar replied:
Failing tests are acos,asin,atan,cos,cosh,exp,log10,log2,log,sin,sih,tan,tanh with CPU & integer types because of "result type Float can't be cast to the desired output type".

I guess `getOpInfo` function will make this simpler

ngimel replied:
Ok, so it's due to inplace integer inputs, cool, thanks for the comment.

ngimel reviewed (commented):
This looks good, I've left minor comments and we can land after that! We'll have to think about how to keep supported dtypes in sync, but it's not for this PR.

crcrpar committed:
bool check for negation

- do bool check after slowpath dispatch
- check only the first tensor of input tensor list

crcrpar committed:
comment: why try-except for inplace fastpath test

ngimel commented:
Can you please rebase to get the signal from CI?

crcrpar commented:
Rebased :)

ngimel commented:
Damn master was in bad state due to forward AD failures, ok, I'll try landing as is. @crcrpar you did run for_each tests locally, right?

crcrpar commented:
@ngimel Yes I ran `test_foreach.py` locally after rebasing but I'll double check

ngimel commented:
It broke the tests unfortunately, will have to revert https://github.com/pytorch/pytorch/runs/2786062691

ngimel commented:
See #59746, please resubmit with a gha skip #59748, then we can remove a skip once kineto is fixed and updated.

crcrpar commented:
sorry about that and thank you for the super fast fix

ngimel commented:
It wasn't your PR's fault.

crcrpar commented:
@ngimel should I resubmit this or will you reland this on my behalf?

ngimel commented:
Relanded already
